### PR TITLE
Add com.github.jms55.Sandbox

### PR DESCRIPTION
**Pull request by JMS55**
Adds an open source sandbox game I made https://github.com/JMS55/sandbox

Does not compile with the inbuilt gameplay recording feature enabled, as I could not figure out how to get the x264enc gstreamer element to work in a flatpak build.

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
bot, build com.github.jms55.sandbox

**Comment by hfiguiere** [comment]
the appdata file doesn't validate.

**Comment by JMS55** [comment]
Should be fixed now.

**Comment by hfiguiere** [comment]
bot, build com.github.jms55.sandbox

**Comment by JMS55** [comment]
Huh, I was testing with appstream-util validate-relax, apparently that didn't catch this. Give me a minute.

**Comment by JMS55** [comment]
I'm not sure what the error is https://github.com/JMS55/sandbox/blob/master/flatpak/com.github.jms55.sandbox.metainfo.xml#L25-L130
```
• style-invalid         : Content before <ul> is too short [14], at least 20 characters required
• style-invalid         : <li> is empty
• style-invalid         : Content before <ul> is too short [4], at least 20 characters required
```

**Comment by bilelmoussaoui** [comment]
bot, build com.github.jms55.sandbox

**Comment by JMS55** [comment]
Btw, should it be `com.github.jms55.Sandbox`, or `com.github.jms55.sandbox`?

**Comment by bilelmoussaoui** [comment]
It seems that the build fails on arm & aarch64, feel free to either fix them or disable them using a flathub.json file. Regarding the appdata, you should validate it using https://github.com/flathub/flathub/wiki/AppData-Guidelines#use-flathubs-appstream-util which is a slightly patched appstream-util to enforce more rules by default.

For the app-id, whatever you want, both are okay. People tend to uppercase the first letter of the latest segment

**Comment by JMS55** [comment]
What does the flathub.json look like? I can't find any examples.

For the appstream data, it seems it has issues with allowing a <code> tag inside a <li>
```
      <li>
        <code>+</code>
        Increases brush size
      </li>
```
And It wants more content before <ul>
Kinda dumb rules, but I'll see if I can fix them.

**Comment by bilelmoussaoui** [comment]
> What does the flathub.json look like? I can't find any examples.

https://github.com/flathub/flathub/wiki/App-Maintenance#flathubjson

> For the appstream data, it seems it has issues with allowing a `<code>` tag inside a `<li>`
> 
> ```
>       <li>
>         <code>+</code>
>         Increases brush size
>       </li>
> ```
> 
> And It wants more content before `<ul>`
> Kinda dumb rules, but I'll see if I can fix them.

code tags are not supported

**Comment by JMS55** [comment]
Both those problems are fixed now.
You probably want to note somewhere or modify the validation tool to say that code tags aren't supported, as the appstream spec says they are https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-description

**Comment by bilelmoussaoui** [comment]
> Both those problems are fixed now.
> 
> You probably want to note somewhere or modify the validation tool to say that code tags aren't supported, as the appstream spec says they are https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-description

The spec itself says they are supported, but appstream-glib doesn't 

bot, build com.github.jms55.sandbox

**Comment by bilelmoussaoui** [comment]
bot, build com.github.jms55.Sandbox

**Comment by JMS55** [comment]
Ah, I didn't rename the manifest in the flathub repo (I did in upstream though), 1s.

**Comment by bilelmoussaoui** [comment]
bot, build com.github.jms55.Sandbox

**Comment by bilelmoussaoui** [comment]
> Adds an open source sandbox game I made https://github.com/JMS55/sandbox
> 
> Does not compile with the inbuilt gameplay recording feature enabled, as I could not figure out how to get the x264enc gstreamer element to work in a flatpak build.

regarding x264, you should be able to add it to your manifest with something similar to https://github.com/flathub/org.avidemux.Avidemux/blob/03a4c02d7ff1326d8890105bcb50892e87c5f038/org.avidemux.Avidemux.yaml#L43

**Comment by JMS55** [comment]
Thank you, please hold off on merging this for now then, let me try that.

File: com.github.jms55.Sandbox.json [hunk 1]
```diff
@@ -0,0 +1,45 @@
+{
+    "app-id": "com.github.jms55.Sandbox",
+    "runtime": "org.freedesktop.Platform",
+    "runtime-version": "19.08",
+    "sdk": "org.freedesktop.Sdk",
+    "sdk-extensions": [
+        "org.freedesktop.Sdk.Extension.rust-stable"
+    ],
+    "finish-args": [
```
Comment (bilelmoussaoui) [inline com.github.jms55.Sandbox.json:9]: You might want `--filesystem=xdg-videos` to be able to save the recorder videos

Comment (JMS55) [reply com.github.jms55.Sandbox.json:9]: Does `--filesystem=xdg-videos/sandbox` work, if I only write to a specific subfolder?

**Comment by JMS55** [comment]
I tried using the following manifest, gstreamer still could not find the x264enc element. Any ideas?
```json
{
    "app-id": "com.github.jms55.Sandbox",
    "runtime": "org.freedesktop.Platform",
    "runtime-version": "19.08",
    "sdk": "org.freedesktop.Sdk",
    "sdk-extensions": [
        "org.freedesktop.Sdk.Extension.rust-stable"
    ],
    "finish-args": [
        "--device=dri",
        "--socket=x11",
        "--share=ipc",
        "--filesystem=xdg-videos/sandbox"
    ],
    "build-options": {
        "append-path": "/usr/lib/sdk/rust-stable/bin",
        "env": {
            "CARGO_HOME": "/run/build/sandbox/cargo"
        }
    },
    "command": "sandbox",
    "modules": [
        {
            "name": "sandbox",
            "buildsystem": "simple",
            "build-commands": [
                "head -n -8 Cargo.toml > t && mv t Cargo.toml",
                "cargo --offline fetch",
                "cargo --offline build --release --features video-recording",
                "install -Dm755 ./target/release/sandbox -t /app/bin/",
                "install -Dm644 ./flatpak/com.github.jms55.Sandbox.metainfo.xml -t /app/share/metainfo/",
                "install -Dm644 ./flatpak/com.github.jms55.Sandbox.desktop -t /app/share/applications/",
                "install -Dm644 ./flatpak/com.github.jms55.Sandbox.svg -t /app/share/icons/hicolor/scalable/apps/"
            ],
            "sources": [
                {
                    "type": "dir",
                    "path": "../"
                },
                "crate-sources.json"
            ]
        },
        {
            "name": "x264",
            "config-opts": [
                "--disable-cli",
                "--enable-shared"
            ],
            "sources": [
                {
                    "type": "archive",
                    "url": "https://code.videolan.org/videolan/x264/-/archive/1771b556ee45207f8711744ccbd5d42a3949b14c/x264-1771b556ee45207f8711744ccbd5d42a3949b14c.tar.bz2",
                    "sha256": "b31c5db5337873b9ee42b8cbc60b56d60ce368fa7e4c4eff90a729b36eeb8326"
                }
            ]
        }
    ]
}
```

**Comment by bilelmoussaoui** [comment]
@JMS55 the shipped gstreamer modules in the freedeskop SDK don't contain everything. Feel free to compile/include whatever you need that's missing in 19.08

File: com.github.jms55.Sandbox.json (hunk 1 above)
Comment (bilelmoussaoui) [reply com.github.jms55.Sandbox.json:9]: Should work fine

**Comment by JMS55** [comment]
I got the video recording feature to work. The only issue is that it broke at some point and only records for half the time, once I fix that I'll push to the flathub repo.

File: com.github.jms55.Sandbox.json (hunk 1 above)
Comment (JMS55) [reply com.github.jms55.Sandbox.json:9]: This did not work actually, only xdg-videos without a subpath worked.

**Comment by JMS55** [comment]
Ok, should be fixed now!

**Comment by hfiguiere** [comment]
bot, build com.github.jms55.Sandbox

**Comment by JMS55** [comment]
Is there anything left I need to do?

**Pull request closed without merging**